cpp/src/binning/bin.cu

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <thrust/binary_search.h>
+#include <thrust/execution_policy.h>
+#include <thrust/functional.h>
+#include <thrust/pair.h>
+#include <cudf/binning/bin.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/span.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <limits>
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+namespace cudf {
+
+// Sentinel used to indicate that an input value should be placed in the null
+// bin.
+// NOTE: In theory if a user decided to specify 2^31 bins this would fail. We
+// could make this an error in Python, but that is such a crazy edge case...
+constexpr size_type NULL_VALUE{std::numeric_limits<size_type>::max()};
+
+namespace detail {
+namespace {
+
+template <typename T, typename LeftComparator, typename RightComparator>
+struct bin_finder {
+  bin_finder(detail::device_span<const T> left_span, detail::device_span<const T> right_span)
+    : m_left_span(left_span), m_right_span(right_span)
+  {
+  }
+
+  __device__ size_type operator()(thrust::pair<T, bool> input_value) const
+  {
+    // Immediately return sentinel for null inputs.
+    if (!input_value.second) return NULL_VALUE;
+
+    T value = input_value.first;
+    auto bound =
+      thrust::lower_bound(thrust::seq, m_left_span.begin(), m_left_span.end(), value, m_left_comp);
+
+    // Exit early and return sentinel for values that lie below the interval.
+    if (bound == m_left_span.begin()) { return NULL_VALUE; }
+
+    // We must subtract 1 because lower bound returns the first index
+    // _greater than_ the value. This is safe because bound == m_left edges
+    // would already have triggered a NULL_VALUE return above, so there's
+    // no risk of negative values or wraparound for -1 here.
+    auto index = bound - m_left_span.begin() - 1;
```

Comment:
When doing iterator manipulation like this, it better reveals your intent to use these functions. Otherwise, when doing a quick read it's not obvious if this is normal arithmetic or pointer manipulation. 
```suggestion
    auto index = thrust::distance(thrust::prev(bound), m_left_span.begin());
```